tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/core/NN.java

```diff
@@ -0,0 +1,379 @@
+   * @param <U> the data type of the logits
+   * @param <T> the number type of the operands
+   * @return the softmax cross entropy loss. Its type is the same as `logits` and its shape is the
+   *     same as `labels` except that it does not have the last dimension of `labels`.
+   */
+  @Endpoint(name = "softmaxCrossEntropyWithLogits")
+  public static <U extends TType, T extends TNumber> Operand<T> softmaxCrossEntropyWithLogits(
+      Scope scope, Operand<T> labels, Operand<U> logits, int axis) {
+    axis = axis % logits.asOutput().shape().numDimensions();
+    if (axis < 0) {
+      axis += logits.asOutput().shape().numDimensions();
+    }
+
+    Operand precise_logits =
+        logits; // cannot use generics cause logits of bool gets cast to TFloat32
+
+    boolean convertToFloat32 =
+        logits.asOutput().dataType() == TFloat16.DTYPE
+            || logits.asOutput().dataType() == TBfloat16.DTYPE;
+    if (convertToFloat32) {
+      precise_logits = Cast.create(scope, logits, TFloat32.DTYPE);
+    }
+    /* cannot use generics on DataType because precis_logits may have been cast. */
+    DataType dtype = precise_logits.asOutput().dataType();
+    labels = Cast.create(scope, labels, dtype);
+    Operand<TInt64> inputRank =
+        Cast.create(scope, Rank.create(scope, precise_logits), TInt64.DTYPE);
+    Shape shape = logits.asOutput().shape();
+
+    // Move the dim to the end if dim is not the last dimension.
+    if (axis != -1 && axis != precise_logits.asOutput().shape().numDimensions() - 1) {
+      precise_logits = moveDimToEnd(scope, precise_logits, axis, inputRank);
+      labels = moveDimToEnd(scope, labels, axis, inputRank);
+    }
+
+    Shape inputShape = precise_logits.asOutput().shape();
+    precise_logits = flattenOuterDims(scope, precise_logits);
+    labels = flattenOuterDims(scope, labels);
+    SoftmaxCrossEntropyWithLogits<T> smax =
+        SoftmaxCrossEntropyWithLogits.create(scope, precise_logits, labels);
+    /* cannot use generic on cost, because cost may be recast later. */
+    Operand cost = smax.loss();
+    Operand<TInt64> outputShape =
+        Slice.create(
+            scope,
+            Constant.vectorOf(scope, inputShape.asArray()),
+            Constant.vectorOf(scope, new long[] {0}),
+            Constant.vectorOf(scope, new long[] {inputShape.numDimensions() - 1}));
+    cost = Reshape.create(scope, cost, outputShape);
+    if (scope.env().isGraph() && !shape.hasUnknownDimension()) {
+      long[] array = shape.asArray();
+      long[] newArray = new long[array.length - 1];
+      if (axis < 0) {
+        axis = shape.numDimensions() + axis;
+      }
+      for (int i = 0; i < axis; i++) {
+        newArray[i] = shape.size(i);
+      }
+      for (int i = axis + 1; i < shape.numDimensions(); i++) {
+        newArray[i - 1] = shape.size(i);
+      }
+      Shape newShape = Shape.of(newArray);
+      cost = Reshape.create(scope, cost, Constant.vectorOf(scope, newShape.asArray()));
```

Comment:
no need to convert `newArray` to a `Shape` if you pass it back as an array in `Constant.vectorOf`.

Reply:
OK